### ISSUE TITLE
`tailscale ssh` doesn't work with the macOS sandbox app

### ISSUE THREAD

**Issue by maisem** [tailscale/tailscale#4518]
### What is the issue?

```bash
$ /Applications/Tailscale.app/Contents/MacOS/Tailscale ssh ws-1
Pseudo-terminal will not be allocated because stdin is not a terminal.
kex_exchange_identification: Connection closed by remote host
Connection closed by UNKNOWN port 65535
```

### Steps to reproduce

Run `/Applications/Tailscale.app/Contents/MacOS/Tailscale ssh <host>` on any macOS machine, where `<host>` is any Tailscale node.

### OS

macOS

### OS version

12.2.1

### Tailscale version

1.24

**Comment by bradfitz** [comment]
It's likely this just won't work in the App Store version.

We might have to detect that version and said, "Sorry, use a different macOS version."

/cc @nickoneill @crawshaw

**Comment by bradfitz** [comment]
As of #4531 it's better, but still not great. See the tcsetattr error and inability to
hit Ctrl-D, for instance:

```
    bradfitz@laptop ~ % tailscale.app ssh foo@bar
    tcsetattr: Operation not permitted
    # Authentication checked with Tailscale SSH.
    # Time since last authentication: 1h13m22s
    foo@bar:~$ ^D
    ^D
    ^D
```

**Comment by mayakacz** [comment]
I think you are saying that macOS won't work as the server you SSH to, but it should still work fine as the client you SSH from. (Or are you saying the client will not work?)

**Comment by bradfitz** [comment]
MacOS with Tailscale wrapper client won't work well. Bad enough to be embarrassing and will invite tons of dup bugs we'll be unable to do anything about.

MacOS with system ssh client is fine. We just don't get the magic host key checking bit yet.

**Comment by mihaip** [comment]
I looked into this a bit. The `tcsetattr: Operation not permitted` is due to not being able to set the tty in raw mode at https://github.com/openssh/openssh-portable/blob/2dc328023f60212cd29504fc05d849133ae47355/sshtty.c#L93. 

I thought that this was maybe due to the ssh process not inheriting some permissions, so I tried doing this directly in the ssh.go CLI command (via `termios.Raw`), but that got the same permission error.

Looking at the console, there is a `Sandbox: ssh(10318) deny(1) file-ioctl /dev/ttys001` line (`tcsetattr` is implemented in terms of `ioctl`), which suggests that this is indeed a sandbox limitation.

Looking at the default sandbox profile for apps (`/System/Library/Sandbox/Profiles/application.sb`), there are a few places where `file-ioctl` is allowed:  `(allow file-read* file-write* file-ioctl (literal "/dev/ptmx"))` and something relating to application groups that I don't think is relevant.

Can we perhaps use `/dev/ptmx` (which can be put into raw mode) and connect it to stdin?

This may also be something for @nickoneill to bring up at the next WWDC. I can also try to make a self-contained sample and see if Quinn or someone else from Apple in their forums has ideas.

**Comment by mihaip** [comment]
Posted https://developer.apple.com/forums/thread/709444 on Apple's forums.

**Comment by manishrjain** [comment]
Echo this same issue. My teammates aren't able to ssh into machines where tailscale is controlling ssh via their Mac systems. So, I had to disable tailscale ssh and revert back to normal ssh.

**Comment by bradfitz** [comment]
@manishrjain, you can just use the normal "ssh" command from macOS to a Linux Tailscale SSH machine. You don't need the "tailscale ssh" subcommand.

**Comment by manishrjain** [comment]
I see. I'll ask my team to try that.

**Comment by skhaz** [comment]
I have the same issue, tailscale installed using brew cask gets stuck forever on "VPN starting...", o logs I have found this

```
sudo dmesg | grep tail
[ 3855.911694]: 1 duplicate report for Sandbox: Tailscale(8639) deny(1) file-read-data /Users/skhaz/Library/Group Containers/W5364U7YZB.group.io.tailscale.ipn.macsysSandbox: contentlinkingd(1415) deny(1) user-preference-read kCFPreferencesAnyApplicationmemorystatus: assertion priority 40 overrides priority 30 for ClassroomSettings:3754
```

**Comment by aaomidi** [comment]
Not sure if the base `ssh` command just works:


```
debug2: service_accept: ssh-userauth
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: tailscale
debug1: No more authentication methods to try.
amir@skyzone: Permission denied (tailscale).
```

Note I can login fine using the web ssh console.

**Comment by geoffeg** [comment]
I'm experiencing the same issue. The built-in (/usr/bin/ssh) and homebrew (/opt/homebrew/bin/ssh) are not able to connect to a server running tailscale ssh. The same error as `aaomidi` reported above. The web ssh console also works fine for me.

**Comment by bradfitz** [comment]
@geoffeg, this issue is specifically about the sandboxed Tailscale CLI which neither of your example ssh paths are using. So this is the wrong issue.

**Comment by evilhamsterman** [comment]
While not a solution for the wrapper subcommand I opened a PR #10186 that will allow the native `ssh` client and related commands to get the host keys from Tailscale

**Comment by lbrichards** [comment]
Ok, the AppStore version is sandboxed and won't allow incoming SSH connections, and likewise won't allow the command `tailscale set --ssh` (which raises the message `The Tailscale SSH server does not run in sandboxed Tailscale GUI builds.` So is this a dead end, or is there an un-sandboxed version that I can use?

**Comment by lbrichards** [comment]
Of course, if I don't mind typing my password I can do like `ssh user@desktop`, but it's more convenient to use key authentication.

**Comment by Nezteb** [comment]
I think the differences in macOS Tailscale SSH functionality should be clarified in the knowledge base a bit, perhaps these pages:
- https://tailscale.com/kb/1065/macos-variants
- https://tailscale.com/kb/1193/tailscale-ssh?q=ssh#limitations

My hypothesis so far:

- Any of the variants will allow SSH auth via regular keys/config while connected to the same tailnet, assuming any receiving macOS machines have "Remote Login" enabled:

    <img width="714" alt="screenshot_2023-12-30_at_23 57 44@2x" src="https://github.com/tailscale/tailscale/assets/3588798/133cd84e-fe1e-4832-9667-fa3f8657c781">

- I think the sandboxed Mac App Store version is the only one that can't use the automatic Tailscale SSH auth (which doesn't require manual keys/config setup).

I suggest this because I was confused about it and only found one search result mentioning it: https://www.reddit.com/r/Tailscale/comments/13cncn7/tailscale_mac_ssh/

😅

**Comment by agottardo** [comment]
The version of Tailscale distributed directly by Tailscale (Standalone variant, distributed on `pkgs.tailscale.com`) is the only one that currently supports this. There is no plan to bring this to the App Store version, as it is technically unfeasible due to Apple's sandboxing restrictions. Please see https://tailscale.com/kb/1065/macos-variants for more details about the different Tailscale variants on macOS.

**Comment by ditsuke** [comment]
> I think "sandboxed macOS builds" is not very obvious, at least not to me (sorry). Initially I expected that only the Tailscale app installed from App Store is "sandboxed", but turns out that the standalone one is sandboxed as well (because tailscale ssh doesn't work).

@agottardo It appears that I and some others get this on the standalone variant as well?

**Comment by viswatejan** [comment]
Yes, I downloaded 1.76.6 directly from the tailscale website and it is still showing the following error when I am enabling the SSH on my M4 Mac.

`The Tailscale SSH server does not run in sandboxed Tailscale GUI builds.`

**Comment by Soviut** [comment]
I also tried Tailscale 1.76.6 and 1.76.1 and both give me the same error.

**Comment by ben-xD** [comment]
Same error here.

Setting up a new machine with tailscale after downloading the one from "Download tailscale" button on https://tailscale.com/kb/1065/macos-variants, I still get this error. 

Trying to set up macOS as an SSH host, over tailscale, to connect from a tailscale client.

**Comment by nalzok** [comment]
I'm getting exactly the same error with the standalone variant (Version 1.78.1) on macOS (Ventura 13.7.2). Please re-open the issue.

**Comment by MitchDresdner** [comment]
Same here Sonoma 14.7.1, running Standalone version after uninstalling App Store version

tailscale set --ssh
2025/01/19 15:48:32 The Tailscale SSH server does not run in sandboxed Tailscale GUI builds.

**Comment by johnnydecimal** [comment]
Same here. I had the MAS version installed. I removed it and re-installed the Tailscale-1.78.1-macos.pkg.

**Comment by EdoAPP** [comment]
Having the same error after trying it from multiple machines using the `standalone` version

**Comment by agottardo** [comment]
There is no actual bug here... but that error message should have been less technical to prevent user confusion. Let me explain here.

The error you're seeing says that the built-in Tailscale SSH server is unable to run in any sandboxed Tailscale build for macOS. Sandboxing limitations by Apple prevent the Tailscale network extension from running as an SSH server: the extension does run as root but is sandboxed by the operating system, so it won't be able to open a terminal on behalf of your local user account. We must break out of the sandbox to do that.

We could fix this in the future by bundling a separate process (a _LaunchDaemon_, more specifically) acting as a SSH server within the Standalone variant. This is a lot of work as it requires an intense amount of XPC between this SSH daemon and the pre-existing system extension. We would also need to create an UI experience to escalate as root in the client UI to install such LaunchDaemon. We will never, ever be able to support this thing on the App Store variant as it is prohibited by the App Store guidelines.

For now, you may run your own SSH server or simply use the built-in SSH server in macOS, and then connect to that server using Tailscale as a workaround.

Advanced users may also install the open-source variant of Tailscale (`brew install tailscale`) which is not subject to restrictions, includes an SSH server, but also comes with a large number of limitations (for starters, it doesn't have a GUI and doesn't integrate with the macOS VPN settings).

I'm re-opening this issue to track bundling an SSH server within the Standalone variant of Tailscale for macOS. But as I explained above, it is a lot of work and I cannot provide any timeline.

**Comment by johnnydecimal** [comment]
@agottardo Super helpful, thanks. If you think that you're the victim of a bug, that's frustrating.

When you know it’s just a limitation, the frustration goes away!

**Comment by Soviut** [comment]
> For now, you may run your own SSH server or simply use the built-in SSH server in macOS, and then connect to that server using Tailscale as a workaround.

Documentation for this would be really helpful.

> Advanced users may also install the open-source variant of Tailscale (`brew install tailscale`) which is not subject to restrictions, includes an SSH server, but also comes with a large number of limitations (for starters, it doesn't have a GUI and doesn't integrate with the macOS VPN settings).

This would also be good to be documented since I'm not sure what "not integrating with the VPN settings" limits me from doing.

The Tailscale Youtube series has been excellent and I could see "escaping sandbox mode on a mac" being a good topic for a video.

**Comment by cjlacz** [comment]
@agottardo I seem to be getting the same error even with the brew version installed.

**Comment by githubbermoon** [comment]
im getting the sandboxing error when trying to ssh through the cli. i have installed the standalone variant .pkg version 1.84.1. help...

**Comment by andreagrandi** [comment]
> im getting the sandboxing error when trying to ssh through the cli. i have installed the standalone variant .pkg version 1.84.1. help...

same, I had to enable MacOS SSH (remote login), which is not ideal

**Comment by douglaspetrin** [comment]
Same problem, unfortunately. So, based on what @agottardo said, we should use an alternative SSH server, enable built-in SSH on macOS, or try using tailscale from Homebrew. 

Edit: maybe this link can solve things by building the binary: https://github.com/tailscale/tailscale/wiki/Tailscaled-on-macOS
However, you will need to set 100.100.100.100 as your DNS server yourself

**Comment by animalrap** [comment]
> [@agottardo](https://github.com/agottardo) I seem to be getting the same error even with the brew version installed.

Same with the standalone. I'm confused because I understand @agottardo response, but regarding the App Store version, not standalone. This definitely seems like a bug for versions outside the App Store.

**Comment by Technophobe01** [comment]
I am unfortunately experiencing this issue on macOS Tahoe 26.2 with Tailscale 1.92.3.

My experience:

- GUI app (standalone, non-App Store): 
   MagicDNS works, but SSH hosting fails with "The Tailscale SSH server does not run in sandboxed Tailscale GUI builds."
- CLI (Homebrew): 
   SSH hosting works via `tailscale set --ssh`, but MagicDNS doesn't work (hostnames don't resolve)

This creates a frustrating situation where macOS users must choose between:

  1. DNS hostname resolution (GUI app)
  2. SSH server hosting (CLI only)

It would be great to have full feature parity on macOS, allowing both MagicDNS and SSH hosting in the same installation.

Happy to help fix or test, or provide additional debugging information if needed. 

IMPORTANT NOTE: The Sandbox driver has to be uninstalled via:

System Preferences -> General - Login Items / Extensions
Look for: Extensions By Category, Select Network Extensions, and delete TailScale Extension... 

Summary

Current state:
  - Official Tailscale GUI app installed and connected
  - MagicDNS should work (can resolve hostnames to SSH to other machines)
  - SSH hosting on this Mac is not available with the GUI app (known limitation)

Workaround for incoming SSH: Enable macOS's built-in SSH server:
  - System Settings → General → Sharing → Remote Login → Enable

This allows SSH via Tailscale IP addresses (e.g., ssh pkjar@100.122.221.2).

**Comment by sefatanam** [comment]
So, I did this in my mac,
* System Settings → General → Sharing → Remote Login → Enable

& then able to connect via `Termius` in ios smoothly via ssh. Thanks a lot @Technophobe01

**Comment by Technophobe01** [comment]
PR #18272 fixes MagicDNS for the macOS CLI/Homebrew version (`tailscaled`). Four issues were identified and fixed:

1. **MatchDomains not set for darwin** — `/etc/resolver/` files were never created for MagicDNS domains
2. **Split DNS routes missing from MatchDomains** — domains with custom resolvers had no resolver files
3. **macOS intercepts port 53** — local DNS listener now uses port 5533 instead
4. **`dns.CleanUp` race condition** (#18800) — startup cleanup was deleting resolver files from running instances

Tested and working on macOS Homebrew `tailscaled` with MagicDNS, short hostnames, and split DNS routes.

**Comment by mbUSC** [comment]
I am running into the same issue with the standalone version


 `tailscale set --ssh` fails with this error:
```
2026/05/10 20:36:23 The Tailscale SSH server does not run in sandboxed Tailscale GUI builds.
```

**Comment by codeninja** [comment]
Also seeing the issue on 1.96.5 in the mac standalone version.

**Comment by convers39** [comment]
> System Settings → General → Sharing → Remote Login → Enable

enabled, same error 

```
2026/05/29 18:00:01 The Tailscale SSH server does not run in sandboxed Tailscale GUI builds.
```

**Comment by Technophobe01** [comment]
For anyone hitting the MagicDNS gap on the **Homebrew/CLI** variant (SSH hosting works, but Tailscale hostnames don't resolve), I'm maintaining an unofficial, build-it-yourself branch that fixes MagicDNS for that variant as a stopgap:

https://github.com/Technophobe01/tailscale/tree/fix/macos-magicdns-homebrew

It's source-only — you build and install it on your own machine (there's a `justfile` and a short guide). **Unofficial, not affiliated with or supported by Tailscale — please don't file issues about this build on this repo; use the fork.**

This isn't a substitute for the maintainers' preferred long-term direction (a LaunchDaemon SSH server bundled in the Standalone variant, talking over XPC to the existing system extension), which addresses the complementary "SSH server in Standalone" side. It just keeps the CLI variant usable for MagicDNS in the meantime, and I'm rebasing it onto current `main` as I go. My thinking is that, while not the long-term direction, it might still be useful to others.

I plan to explore that LaunchDaemon/XPC direction separately, interested to see if that can be built. It does feel like the better solution.

**Comment by insidevoid** [comment]
> System Settings → General → Sharing → Remote Login → Enable

enabled, but still getting the same error... has anyone found solution to this?

2026/09/15 19:13:29 The Tailscale SSH server does not run in sandboxed Tailscale GUI builds.
